File: src/types/channel.ts
import { Value } from "./values";
import { LiveMap } from "./values";
import { PathObject } from "./pathobjects";

// The type returned by `client.channels.get()` in the Ably Realtime client
export interface Channel {
  object: {
    get<T extends Record<string, Value>>(): Promise<PathObject<LiveMap<T>>>;
    on(event: "syncing" | "synced", callback: () => void): void;
    off(event: "syncing" | "synced", callback?: Function): void;
  };
}

File: swift/Sources/AblyLiveObjects/channel.swift
// The type returned by `client.channels.get()` in the Ably Realtime client
public protocol Channel: Sendable {
    var object: RealtimeObjectProtocol { get }
}

// Lawrence: Added for Swift; this is what's returned by `channel.object`.
public protocol RealtimeObjectProtocol {
    // TODO: is `get` a great name for this? It's a keyword in Swift in certain circumstances
    func get() async throws -> any LiveMapPathObject
    func on(_ event: RealtimeObjectEvent, callback: @escaping () -> Void)
}

public enum RealtimeObjectEvent {
    case syncing
    case synced
}
